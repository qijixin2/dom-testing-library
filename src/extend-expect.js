import extensions from './jest-extensions'

const {toBeInTheDOM, toHaveTextContent, toHaveAttribute} = extensions
expect.extend({toBeInTheDOM, toHaveTextContent, toHaveAttribute})
