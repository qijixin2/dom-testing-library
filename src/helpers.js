const globalObj = typeof window === 'undefined' ? global : window
// Constant node.nodeType for text nodes, see:
// https://developer.mozilla.org/en-US/docs/Web/API/Node/nodeType#Node_type_constants
const TEXT_NODE = 3

// Currently this fn only supports jest timers, but it could support other test runners in the future.
function runWithRealTimers(callback) {
  return _runWithRealTimers(callback).callbackReturnValue
}

function _runWithRealTimers(callback) {
  const timerAPI = {
    clearImmediate,
    clearInterval,
    clearTimeout,
    setImmediate,
    setInterval,
    setTimeout,
  }

  // istanbul ignore else
  if (typeof jest !== 'undefined') {
    jest.useRealTimers()
  }

  const callbackReturnValue = callback()

  const usedJestFakeTimers = Object.entries(timerAPI).some(
    ([name, func]) => func !== globalObj[name],
  )

  if (usedJestFakeTimers) {
    jest.useFakeTimers(timerAPI.setTimeout?.clock ? 'modern' : 'legacy')
  }

  return {
    callbackReturnValue,
    usedJestFakeTimers,
  }
}

const jestFakeTimersAreEnabled = () =>
  Boolean(_runWithRealTimers(() => {}).usedJestFakeTimers)

// we only run our tests in node, and setImmediate is supported in node.
// istanbul ignore next
function setImmediatePolyfill(fn) {
  return globalObj.setTimeout(fn, 0)
}

function getTimeFunctions() {
  // istanbul ignore next
  return {
    clearTimeoutFn: globalObj.clearTimeout,
    setImmediateFn: globalObj.setImmediate || setImmediatePolyfill,
    setTimeoutFn: globalObj.setTimeout,
  }
}

const {clearTimeoutFn, setImmediateFn, setTimeoutFn} = runWithRealTimers(
  getTimeFunctions,
)

function getDocument() {
  /* istanbul ignore if */
  if (typeof window === 'undefined') {
    throw new Error('Could not find default container')
  }
  return window.document
}
function getWindowFromNode(node) {
  if (node.defaultView) {
    // node is document
    return node.defaultView
  } else if (node.ownerDocument && node.ownerDocument.defaultView) {
    // node is a DOM node
    return node.ownerDocument.defaultView
  } else if (node.window) {
    // node is window
    return node.window
  } else if (node.then instanceof Function) {
    throw new Error(
      `It looks like you passed a Promise object instead of a DOM node. Did you do something like \`fireEvent.click(screen.findBy...\` when you meant to use a \`getBy\` query \`fireEvent.click(screen.getBy...\`, or await the findBy query \`fireEvent.click(await screen.findBy...\`?`,
    )
  } else {
    // The user passed something unusual to a calling function
    throw new Error(
      `Unable to find the "window" object for the given node. Please file an issue with the code that's causing you to see this error: https://github.com/testing-library/dom-testing-library/issues/new`,
    )
  }
}

function checkContainerType(container) {
  if (
    !container ||
    !(typeof container.querySelector === 'function') ||
    !(typeof container.querySelectorAll === 'function')
  ) {
    throw new TypeError(
      `Expected container to be an Element, a Document or a DocumentFragment but got ${getTypeName(
        container,
      )}.`,
    )
  }

  function getTypeName(object) {
    if (typeof object === 'object') {
      return object === null ? 'null' : object.constructor.name
    }
    return typeof object
  }
}

export {
  getWindowFromNode,
  getDocument,
  clearTimeoutFn as clearTimeout,
  setImmediateFn as setImmediate,
  setTimeoutFn as setTimeout,
  runWithRealTimers,
  checkContainerType,
  jestFakeTimersAreEnabled,
  TEXT_NODE,
}
