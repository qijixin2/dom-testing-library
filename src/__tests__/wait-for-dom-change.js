import {waitForDomChange} from '../'
// adds special assertions like toBeTruthy
import 'jest-dom/extend-expect'
import {render} from './helpers/test-utils'

const skipSomeTime = delayMs =>
  new Promise(resolve => setTimeout(resolve, delayMs))

// Using `setTimeout` >30ms instead of `wait` here
// because `mutationobserver-shim` uses `setTimeout` ~30ms.
const skipSomeTimeForMutationObserver = (delayMs = 50) =>
  skipSomeTime(delayMs, 50)

test('it waits for the next DOM mutation', async () => {
  const successHandler = jest.fn().mockName('successHandler')
  const errorHandler = jest.fn().mockName('errorHandler')

  waitForDomChange().then(successHandler, errorHandler)

  // Promise callbacks are always asynchronous.
  expect(successHandler).toHaveBeenCalledTimes(0)
  expect(errorHandler).toHaveBeenCalledTimes(0)

  await skipSomeTimeForMutationObserver()

  // No more expected calls without DOM mutations.
  expect(successHandler).toHaveBeenCalledTimes(0)
  expect(errorHandler).toHaveBeenCalledTimes(0)

  document.body.appendChild(document.createElement('div'))

  await skipSomeTimeForMutationObserver()

  expect(successHandler).toHaveBeenCalledTimes(1)
  expect(successHandler.mock.calls[0]).toMatchSnapshot()
  expect(errorHandler).toHaveBeenCalledTimes(0)
})

test('it waits characterData mutation', async () => {
  const {container} = render(`<div>initial text</div>`)

  const successHandler = jest.fn().mockName('successHandler')
  const errorHandler = jest.fn().mockName('errorHandler')

  waitForDomChange({container}).then(successHandler, errorHandler)

  await skipSomeTimeForMutationObserver()

  expect(successHandler).toHaveBeenCalledTimes(0)
  expect(errorHandler).toHaveBeenCalledTimes(0)

  container.querySelector('div').innerHTML = 'new text'
  await skipSomeTimeForMutationObserver()

  expect(successHandler).toHaveBeenCalledTimes(1)
  expect(errorHandler).toHaveBeenCalledTimes(0)
})

test('it waits for the attributes mutation', async () => {
  const {container} = render(``)

  const successHandler = jest.fn().mockName('successHandler')
  const errorHandler = jest.fn().mockName('errorHandler')

  waitForDomChange({
    container,
  }).then(successHandler, errorHandler)

  await skipSomeTimeForMutationObserver()

  expect(successHandler).toHaveBeenCalledTimes(0)
  expect(errorHandler).toHaveBeenCalledTimes(0)

  container.setAttribute('data-test-attribute', 'PASSED')
  await skipSomeTimeForMutationObserver()

  expect(successHandler).toHaveBeenCalledTimes(1)
  expect(successHandler.mock.calls[0]).toMatchSnapshot()
})

test('it throws if timeout is exceeded', async () => {
  render('')
  const successHandler = jest.fn().mockName('successHandler')
  const errorHandler = jest.fn().mockName('errorHandler')

  waitForDomChange({timeout: 70}).then(successHandler, errorHandler)

  await skipSomeTimeForMutationObserver(100)

  expect(successHandler).toHaveBeenCalledTimes(0)
  expect(errorHandler).toHaveBeenCalledTimes(1)
  expect(errorHandler.mock.calls[0]).toMatchInlineSnapshot(`
Array [
  [Error: Timed out in waitForDomChange.],
]
`)
})

test('does not get into infinite setTimeout loop after MutationObserver notification', async () => {
  const {container} = render(`<div data-testid="initial-element"></div>`)

  const successHandler = jest.fn().mockName('successHandler')
  const errorHandler = jest.fn().mockName('errorHandler')
  jest.useFakeTimers()

  const promise = waitForDomChange({
    container,
    timeout: 70,
    mutationObserverOptions: {attributes: true},
  }).then(successHandler, errorHandler)

  // Expect 2 timeouts to be scheduled:
  // - waitForDomChange timeout
  // - MutationObserver timeout
  expect(setTimeout).toHaveBeenCalledTimes(2)

  expect(successHandler).toHaveBeenCalledTimes(0)
  expect(errorHandler).toHaveBeenCalledTimes(0)

  // Make a mutation
  container.setAttribute('data-test-attribute', 'something changed')

  // Advance timer to expire MutationObserver timeout
  jest.advanceTimersByTime(50)
  jest.runAllImmediates()
  await promise
  expect(setTimeout).toHaveBeenCalledTimes(3)

  // Expect successHandler to be called
  expect(successHandler).toHaveBeenCalledTimes(1)
  expect(errorHandler).toHaveBeenCalledTimes(0)

  // Expect no more setTimeout calls
  jest.advanceTimersByTime(100)
  expect(setTimeout).toHaveBeenCalledTimes(3)
  jest.useRealTimers()
})
