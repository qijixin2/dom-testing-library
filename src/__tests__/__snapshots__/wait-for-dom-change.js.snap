// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`it waits for the attributes mutation 1`] = `
Array [
  Array [
    Object {
      "addedNodes": Array [],
      "attributeName": "data-test-attribute",
      "attributeNamespace": null,
      "nextSibling": null,
      "oldValue": null,
      "previousSibling": null,
      "removedNodes": Array [],
      "target": <div
        data-test-attribute="PASSED"
      />,
      "type": "attributes",
    },
  ],
]
`;

exports[`it waits for the next DOM mutation 1`] = `
Array [
  Array [
    Object {
      "addedNodes": Array [
        <div />,
      ],
      "attributeName": null,
      "attributeNamespace": null,
      "nextSibling": null,
      "oldValue": null,
      "previousSibling": null,
      "removedNodes": Array [],
      "target": <body>
        <div />
      </body>,
      "type": "childList",
    },
  ],
]
`;
